### ISSUE TITLE
Rethinking centerlines.

### ISSUE THREAD

**Issue by vvoovv** [prochitecture/blosm#57]
Transitions on Karl-Marx-Allee in Berlin between way-sections do not look realistic. The original OSM way for the lower part of the cluster is straight.

The lanes should follow the course of the original OSM-way.

It there is an increase in the number of lanes, the lanes in the section with the increased number of lanes are not symmetrical relative to the centerline.

The data structure for the way-sections needs rethinking. 

![image](https://user-images.githubusercontent.com/613295/229095314-dda6526f-7a76-41ef-8080-4e9457095c81.png)

**Comment by polarkernel** [comment]
> The data structure for the way-sections needs rethinking.

This problem has already a long history. The reason for this issue are turn-lanes. I [mentioned](https://github.com/prochitecture/blosm/issues/41#issuecomment-1117094829) it already in May 2022. In a first [proposal](https://github.com/prochitecture/blosm/issues/44#issuecomment-1246806206) for the interface to the cluster data, I suggested to use two values for width, `leftWidth `and `rightWidth`, so that turning lanes can be represented, where the left width may be larger than the right one.

It is not possible to derive this from the number of lanes. After a [short discussion](https://github.com/prochitecture/blosm/issues/44#issuecomment-1263548046), we decided to use only one value for the number of lanes in the case of one-way carriageways. It is required to have the `leftWidth `and `rightWidth` data at the time when the intersection area is constructed, so that its connectors are correctly aligned. See also [here](https://github.com/prochitecture/blosm/issues/44#issuecomment-1249316428). 

After I [mentioned](https://github.com/prochitecture/blosm/issues/44#issuecomment-1286616827) that it's not that easy to find the left and right widths from the tag `turn:lanes`, we finally decided to [drop turn-lanes](https://github.com/prochitecture/blosm/issues/44#issuecomment-1286948755) for the first iteration of the code, which seems to be over now. Afterwards, I [bragged](https://github.com/prochitecture/blosm/issues/44#issuecomment-1287687912) that I could turn it back on in five minutes. Well, that was a long time ago, I'm not sure if that's still true.

I don't know much about GN's yet, but my understanding is that in order to look natural, that kind of transition also needs to be smooth, not rendered as a step. Can you already say what the requirements will be from the Blender side?

**Comment by vvoovv** [comment]
> Can you already say what the requirements will be from the Blender side?

I will describe below how streets are generated now.

![image](https://user-images.githubusercontent.com/613295/229277737-fa40919a-55d5-470c-ba6b-bf7d07e91036.png)

Suppose we have a polyline curve. A modifier is applied to the curve to shift it and/or trim it from one or both ends. A modifier may generate mesh out of the given curve and pass the input curve to the output, so another modifier can be applied to the curve. That can be represented as a tree of modifiers.

So we start from the red curve on the image above. Than apply a modifier to shift the red curve to the curve 1A to generate the upper roadway later. Another modifier is applied to the red curve to shift it to the curve 1B to generate the lower roadway later. A similar modifier creates the curve 1C to generate a separator between the roadways.

![image](https://user-images.githubusercontent.com/613295/229286765-7698657a-253a-4c39-9473-aca0551c2e28.png)

____
![image](https://user-images.githubusercontent.com/613295/229290077-91443bb4-b88b-4bfa-8385-ca8ca5d17fca.png)

The next set of modifiers split the curve 1A into curves 2A, 2B, 2C.

A profile with the related texture is applied to the curves 2A, 2B, 2C to create the roadways with asphalt surface and road markings. In addition, the curve 2C is offsetted to apply a symmetrical profile.

A transition mesh is added to the roadway that originates from the curve 2B.

**Comment by polarkernel** [comment]
Some questions for my understanding.

> The next set of modifiers split the curve 1A into curves 2A, 2B, 2C.

- In the current implementation, the curve 2A is separated from the others by the transition area between the change of lane numbers, or do you skip this transition area?
- Is the split between 2B and 2C in the middle of the remaining length?

> A profile with the related texture is applied to the curves 2A, 2B, 2C to create the roadways with asphalt surface and road markings.

- Do the curves 2A and 2B have 3 lanes and 4 lanes for the curve 2C ?

> In addition, the curve 2C is offsetted to apply a symmetrical profile.

- Is the offset for the curve 2C half of a lane width?

> A transition mesh is added to the roadway that originates from the curve 2B.

- Am I right when I assume that this transition mesh is the part that makes the transition smooth?

**Comment by vvoovv** [comment]
> * In the current implementation, the curve 2A is separated from the others by the transition area between the change of lane numbers, or do you skip this transition area?

The transition area is not skipped at the moment as shown on the screenshot in the first [message](#issue-1649255714).

> * Is the split between 2B and 2C in the middle of the remaining length?

No. The curve 2C defines a transition area for the smooth transition between the segments with different number of lanes. I may have a fixed length.

> * Do the curves 2A and 2B have 3 lanes and 4 lanes for the curve 2C ?

The curve 2A has 3 lanes. The curve 2C has 4 lanes. The curve 2B has 3 lanes and the transition area.

> * Is the offset for the curve 2C half of a lane width?

Yes.

> * Am I right when I assume that this transition mesh is the part that makes the transition smooth?

Yes

**Comment by polarkernel** [comment]
Thanks, now it is clear to me. Fine that these tasks may be solved within Blender. So, let's continue rethinking. The following now needs to be considered:

- There exist changes in the number of lanes that are not due to turn lanes. An example are the ways [4616518](https://www.openstreetmap.org/way/4616518) and [824127300](https://www.openstreetmap.org/way/824127300) along the Lichtenberger Straße (from one lane to two).
- There may be left or right turn lanes, and also turn lanes to both sides ([example](https://www.openstreetmap.org/way/316869464)). It is not enough to know the change in the number of lanes, one also needs their direction.
- Way-clusters with turn lanes have a narrow end and a wide end. The `WayCluster` must record, somehow, which end is which. The way clusters (or their centerlines) have no fixed direction relative to the turn lanes (due to the heuristics used to detect them). 
- The intersection/transition connectors on both sides must be adjusted. The wide side needs to be offset (by half a lane, as in your example at the end of 2C). In theory, the narrow side would not need a transition because the number of lanes does not change. However, since the OSM operators do not draw the lanes with constant spacing, a width transition may be needed. So I suggest keeping this transition area, but the width of the connector needs to be shortened.
- Ways that are not tagged as `oneway` may also have turn lanes.
- How do we get the required information, when the tag `turn:lanes` is not reliable? (See [here](https://github.com/prochitecture/blosm/issues/44#issuecomment-1286948755)).

I propose that we extend the list of arguments of the class `TrimmedWaySection` to transfer the required information.

**Comment by vvoovv** [comment]
> * There may be left or right turn lanes, and also turn lanes to both sides ([example](https://www.openstreetmap.org/way/316869464)). 

The way [316869464](https://www.openstreetmap.org/way/316869464)  has 3 lanes. The previous way [44619069](https://www.openstreetmap.org/way/44619069) has 2 lanes. How can it have lanes to both sides?

**Comment by vvoovv** [comment]
> I propose that we extend the list of arguments of the class `TrimmedWaySection` to transfer the required information.

To trim a curve, the trim lengths at the start and at the end are needed.

**Comment by polarkernel** [comment]
> Who can it have lanes to both sides?

It is tagged as:

```
lanes=3
turn:lanes=left;through|through|right
```
I don't know if in this case the `through` tag overrides the `left` tag, so that the two lanes of the previous segment continue as `through` lanes, while a turn lane is added on the right side. After a somewhat more thorough search, I found that in almost all cases, a second turn lane is associated with the `through` tag. I had to search through all the intersections I know of for quite a while to finally find one case where there is a turn lane on both, the right and left side. In Hamburg, the way [44833752](https://www.openstreetmap.org/way/44833752) expands from 3 forward lanes to the way [547724854](https://www.openstreetmap.org/way/547724854), which then has 4 lanes, but is tagged `turn:lanes=left|through|through|right`.

In my native solution, such tags simply cancelled, and the widening was realized by a transition area.

Should we eventually integrate chatGPT into the addon? :)

**Comment by polarkernel** [comment]
There is another special case, for example way [31975505](https://www.openstreetmap.org/way/31975505), which is tagged `turn:lane=slight_left|slight_left`. But it does not have turn lanes in the sense we are discussing here.

**Comment by vvoovv** [comment]
> * There may be left or right turn lanes, and also turn lanes to both sides ([example](https://www.openstreetmap.org/way/316869464)). It is not enough to know the change in the number of lanes, one also needs their direction.

I probably misunderstood this sentence. I thought you meant additional lanes for turning to the left and to the right from the main course of the roadway.

**Comment by polarkernel** [comment]
> I probably misunderstood this sentence.

I don't think so, your comment was good. I was not aware that the way in my example has only three lanes.

If I understand correctly, you only need a note when a transition is required, as described in your original post [here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1492967362). Additionally, its side (left or right) and its start relative to the start of its centerline is required. It should be sufficient to assume that its width is the width of a lane.

It's not clear to me yet how to do this for streets that are not one-way. More thought is needed there. Also, there are a lot of tags and tag combinations around lanes in OSM. My initial implementation from last year is too primitive to get consistent results.

**Comment by vvoovv** [comment]
A graphical illustration to my previous [message](#issuecomment-1494392657). I thought you meant originally something like this:

![image](https://user-images.githubusercontent.com/613295/229557814-baf9c057-33e2-43b1-bb36-81129670da3a.png)

**Comment by polarkernel** [comment]
> I thought you meant originally something like this:

That's exactly what I meant, but I didn't check the number of lanes in my example, which is only 3. One question, though: For such a case as you illustrate here, would you want to use transitions (like curves 2B and 2C) on both sides, or would it be enough to use just one simple transition area, as I have done so far?

Great graphics! How did you create them? Could be very helpful to understand.

**Comment by vvoovv** [comment]
> That's exactly what I meant, but I didn't check the number of lanes in my example, which is only 3. One question, though: For such a case as you illustrate here, would you want to use transitions (like curves 2B and 2C) on both sides, or would it be enough to use just one simple transition area, as I have done so far?

I'd prefer that transitions will be generated in Blender (namely in Geometry Nodes) provided that the need parameters are given: if an extra lane it to the left or to the right.

Regarding the ambiguity where to place the extra lane. Later a GUI control can be implemented to change the position of the extra lane.

Another option would be to hard code the markup of a roadway in PML. The problem is how to write the condition for a specific street section between the two intersections.

**Comment by vvoovv** [comment]
> Great graphics! How did you create them? Could be very helpful to understand.

I started from the [image](https://commons.wikimedia.org/wiki/File:Fahrbahnmarkierung.png) from the Wikimedia Commons. Then I copied, flipped and pasted part of the image to create the resulting one.

Currently the whole roadway with a texture for the whole roadway is created out of a centerline curve. Alternatively a curve for each lane can be created out of the centerline curve. In the latter case a texture is applied to each lane rather than the whole roadway. I can create a number of modifiers for each type of lanes. Then the needed combination of lanes can be constructed by applying a set of modifiers to the original centerline curve.

**Comment by polarkernel** [comment]
Some more results of further investigations:

It is relatively easy to find changing way widths due to changing number of lanes. Each node in the way-section network of order two (two outgoing ways) which has a different number of lanes on its two sides, marks such an event. The position of the node in relation to the ways also gives the position of the smaller part of a transition.

The direction of the centerline of the way in `TrimmedWaySection` is the direction it was originally drawn by the OSM volunteer operator (at least if I didn't make a mistake somewhere). This will give the direction of one-ways (if the operator observed the rules) and will also give the direction of the markers to be painted on the roadway, should you wish to use them.

A smart decoder will be needed to get the direction of the lane widening based on the `turn:lanes` tags (for a while I even considered using a parser, like for PML). Most cases are obvious, like these (the first increases to the left, the second to the right):
```
lanes turn:lanes
  3   -
  4   left|through|none|none
lanes turn:lanes
  3   -
  4   none|through|through|right
```
Others are ambiguous, and some heuristics has to be used (turn to right?):
```
lanes turn:lanes
  3   left|none|none
  4   left|through|through|right
```
Some cases where lanes are merged are not yet clear to me:
```
lanes turn:lanes
  2   none|merge_to_left
  3   -
lanes turn:lanes
  2   -
  3   none|none|merge_to_left
```
I assume that we need to treat the following cases:
1. Transitions that add a right or left turn lane.
2. Transitions that subtract (merge) a lane at right or at left.
3. Transitions that only change the number of lanes by one (symmetrically).

I never found a case, where the change was more than one lane. I tried to write a query for overpass turbo to verify this, but I was unable to do so. The above cases need to be handled for one-way and for two-way streets.

All necessary information can be generated in `StreetGenerator`. Since you prefer to generate the transitions in Blender, I suggest that we keep the current transition areas (which are instances of the class `IntersectionArea`), so that the data structures do not need to be changed. But their length will then be made very small (1 mm, for example), so that they will be almost invisible and we will get the full beauty of the Geometry Nodes.

I don't know if you want to paint markers on the lanes. If you do, then the necessary information could either be provided additionally (extracted from `StreetGenerator's` tag processor), or you could read the tags yourself.

**Comment by polarkernel** [comment]
Two UPDATES to the last post.

I forgot that the centerlines of the two ways are sometimes not collinear at transitions. The version with the small length then becomes something like a small triangle, so that its connectors on both sides remain perpendicular to the way centerlines.

Your idea to add a transition mesh (curve 2B in your example) may not work for symmetrical transitions, because the change is only half of a lane on each side. I found cases where the number of lanes increased from 1-to-2 up to 3-to-4 lanes. I don't know if you could use some kind of templates for these cases.

**Comment by vvoovv** [comment]
> (for a while I even considered using a parser, like for PML). 

Why is a complex parser needed to parse the values of the tag `turn:lanes`? The values of the turn lanes are simply separated by the pipe symbol '|'.

>   2   none|merge_to_left
>   3   -

There must be an additional way-segment with lanes=1, otherwise it does not make a sense.


> I assume that we need to treat the following cases:
> 
> 1. Transitions that add a right or left turn lane.
> 2. Transitions that subtract (merge) a lane at right or at left.
> 3. Transitions that only change the number of lanes by one (symmetrically).

Agreed.


 
> All necessary information can be generated in `StreetGenerator`. Since you prefer to generate the transitions in Blender, I suggest that we keep the current transition areas (which are instances of the class `IntersectionArea`), so that the data structures do not need to be changed. But their length will then be made very small (1 mm, for example), so that they will be almost invisible and we will get the full beauty of the Geometry Nodes.

Perhaps the transitions areas can be marked as transitions and skipped if necessary.

> I don't know if you want to paint markers on the lanes. If you do, then the necessary information could either be provided additionally (extracted from `StreetGenerator's` tag processor), or you could read the tags yourself.

Yes, I eventually want to paint markers but it will be done on a later stage.

> I forgot that the centerlines of the two ways are sometimes not collinear at transitions. The version with the small length then becomes something like a small triangle, so that its connectors on both sides remain perpendicular to the way centerlines.

Could you please illustrate this case?

> Your idea to add a transition mesh (curve 2B in your example) may not work for symmetrical transitions, because the change is only half of a lane on each side. I found cases where the number of lanes increased from 1-to-2 up to 3-to-4 lanes. I don't know if you could use some kind of templates for these cases.

It seems that a transition polygon will be required for a symmetrical transition.

Can you provide the links to the transitions from 1-2 to 3-4 lanes?

**Comment by vvoovv** [comment]
> Should we eventually integrate chatGPT into the addon? :)

What task could it potentially solve? I am afraid that chatGPT  is not yet aware of our domain.

**Comment by polarkernel** [comment]
> Why is a complex parser needed to parse the values of the tag turn:lanes? The values of the turn lanes are simply separated by the pipe symbol '|'.

I try to keep my sense of humor. Otherwise, I would have despaired of this road generation long ago. :) Finding the tokens is easy, indeed, but their combination to find the additional turn direction can be quite tricky.

> There must be an additional way-segment with lanes=1, otherwise it does not make a sense.

But it isn't. Both cases are from _berlin_karl_marx_allee.osm_ and both change from 2 lanes to 3. The first case occurs between [284760972](https://www.openstreetmap.org/way/284760972) and [4615594](https://www.openstreetmap.org/way/4615594):
```
lanes turn:lanes
  2   none|merge_to_left
  3   -
```
The other one between [281095736](https://www.openstreetmap.org/way/281095736) and [284736594](https://www.openstreetmap.org/way/284736594):
```
lanes turn:lanes
  2   -
  3   none|none|merge_to_left
```
One merges in the smaller part and the other in the wide. OpenStreetMap peculiarity?

> Perhaps the transitions areas can be marked as transitions and skipped if necessary.

Transition areas are constructed separately from the intersection areas. When required, they can even be stored in a separate class.

> Could you please illustrate this case?

<img src="https://user-images.githubusercontent.com/48180714/230622594-b7f067c6-61af-417e-949c-416f690c3c7b.png" width="400">

Sorry, should read 'transition area' instead of 'intersection area'.

> It seems that a transition polygon will be required for a symmetrical transition.

OK.

> Can you provide the links to the transitions from 1-2 to 3-4 lanes?

I probably wrote that in a misleading way. All the transition I have found add no more than one way. I wanted to say that I have found cases from 1 to 2, from 2 to 3, and from 3 to 4 lanes. If you still want example links, I can search for them.

**Comment by polarkernel** [comment]
> What task could it potentially solve? I am afraid that chatGPT is not yet aware of our domain.

If I write this :) sign, please consider it to be a joke. But a joke can also be real. Some times ago, as a test, I asked chatGPT how to construct intersecting surfaces from centerlines. I got a nicely written suggestion to look for the intersection points of the road edges with the intersection surface. I never would have had that in mind!

**Comment by vvoovv** [comment]
> If I write this :) sign, please consider it to be a joke.

The technology is progressing very rapidly. No more jokes. It's already aware of the Blender-OSM addon :)

**Comment by vvoovv** [comment]
> The other one between [281095736](https://www.openstreetmap.org/way/281095736) and [284736594](https://www.openstreetmap.org/way/284736594):
> 
> ```
> lanes turn:lanes
>   2   -
>   3   none|none|merge_to_left
> ```

[284736594](https://www.openstreetmap.org/way/284736594) has `turn:lanes`=`through|through|right` in the current version of the OSM database which is the correct value.

**Comment by polarkernel** [comment]
While working on the evaluation of turn way tags, I found the following special case (just an example of many similar cases).
A first way (assumed to have a forward and a backward lane):
```
id=760490441
lanes=2
oneway=no
```
is followed by a way that adds a turn-lane to the left:
```
id=760490440
lanes:backward=1
lanes:forward=2
lanes=3
oneway=no
turn:lanes:forward=left|through
```
Do you agree, that for such bidirectional ways that add a turn way to the left, we just use a common transition and not a transition mesh?

**Comment by polarkernel** [comment]
As an intermediate step, I have finished the detection of turn-lanes, without using a parser. :) I developed it outside of blosm to simplify testing. As a result, the addition of a lane on the left or right of the way (turn-lane) or the subtraction of a lane on the left or right of the way (merge-lane) is found. It seems to handle all cases of tag combinations, although I had this impression several times during the development and was then surprised by a strange new combination. At least, the framework allows the integration of new cases, should I find more.

Just to illustrate, two cases of such surprises.  For the first one, only the minimum number of lane tags was present, while the other had to be evaluated. The second way must have three forward lanes and therefore two backward lanes. The extra lane is guessed to be to the right, since all the forward lanes point to the right.
<table>
  <tr>
    <td  ALIGN=LEFT>
Way 1:<br>
-----<br>
    "id": 420358888,<br>
    "lanes": "4",<br>
    "oneway": "no",<br><br>
   </td>
<td  ALIGN=LEFT>
Way 2:<br>
-----<br>
  "id": 420358886,<br>
    "lanes": "5",<br>
    "oneway": "no",<br>
    "turn:lanes:forward": "slight_right|slight_right|right"<br>
</td>
<td  ALIGN=LEFT>
<img src="https://user-images.githubusercontent.com/48180714/232226999-222431fd-0525-426e-897f-65088439d6a6.jpg" width="300">
</td>  </tr>
</table>
The second case gives more information. Although there are two lanes pointing to the right for the second way, the added way is guessed to be the `through` lane on the left, because of the rest of the lanes match the input lanes of the first way.
<table>
  <tr>
    <td  ALIGN=LEFT>
Way 1:<br>
-----<br>
    "id": 807311418,<br>
    "lanes": "3",<br>
    "oneway": "yes",<br>
    "turn:lanes": "through|right|right"<br>
   </td>
<td  ALIGN=LEFT>
Way 2:<br>
-----<br>
  "id": 1050717222,<br>
    "lanes": "4",<br>
    "oneway": "yes",<br>
    "turn:lanes": "through|through|right|right"<br>
</td>
<td  ALIGN=LEFT>
<img src="https://user-images.githubusercontent.com/48180714/232225710-21b17a5b-e5fc-4183-a192-dc884500cc3d.png" width="200">
</td>
  </tr>
</table>

**Comment by polarkernel** [comment]
From the detection of the turn-lanes, the left and right widths of the ways at their start and their end have to be calculated so that the intersection or transition areas can be constructed, and their connectors set to the right place. The question now is how to pass this information to Blender. I suggest extending the attributes of the `TrimmedWaySection` class, but I don't know which method you prefer.

**Comment by polarkernel** [comment]
Another ambiguous case. The turn lane detection can't know the side of the extra lanes from these tags:
<table>
  <tr>
    <td  ALIGN=LEFT>
Way 1:<br>
-----<br>
    "id": 662064493,<br>
    "lanes": "3",<br>
    "lanes:backward": "1",<br>
    "lanes:forward": "2",<br>
    "oneway": "no",<br>
    "turn:lanes:forward": "through|through"
   </td>
<td  ALIGN=LEFT>
Way 2:<br>
-----<br>
  "id": 662064487,<br>
    "lanes": "5",<br>
    "lanes:backward": "1",<br>
    "lanes:forward": "4",<br>
    "oneway": "no",<br>
    "turn:lanes:forward": "through|through|through|through"
</td>
</tr>
</table>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Way 2, &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Way 1
<img src="https://user-images.githubusercontent.com/48180714/232314872-c3405e1d-75ee-4547-9639-abe3c8cbd883.jpg" width="500">

**Comment by vvoovv** [comment]
> ```
> id=760490441
> lanes=2
> oneway=no
> ```
> 
> is followed by a way that adds a turn-lane to the left:
> 
> ```
> id=760490440
> lanes:backward=1
> lanes:forward=2
> lanes=3
> oneway=no
> turn:lanes:forward=left|through
> ```
> 
> Do you agree, that for such bidirectional ways that add a turn way to the left, we just use a common transition and not a transition mesh?

That setting would lead to the street section on the image below. Correct? But is it possible in the reality? If a car goes straight on, it does not need to switch to an additional lane to the right.

![image](https://user-images.githubusercontent.com/613295/232329072-bea7e44f-e955-45d6-b624-3403aa490d61.png)

**Comment by polarkernel** [comment]
> That setting would lead to the street section on the image below.

That was the reason for my question. I was not very precise with the terms. With 'common transition' I meant a transition area as it is also used when the number of lanes changes, but without turn lanes (see left image below). With 'transition mesh' I meant a turn lane, constructed in Blender, and also shown in your post. In reality, the case of a left turn in a bidirectional way usually just leads to a widening of the street, where both forward and backward lanes have to make slight curves (see an example in the right image). Therefore, my suggestion.

<img src="https://user-images.githubusercontent.com/48180714/232427088-0a28fee2-aa21-4c58-971e-945a75a332b1.jpg" width="300"> <img src="https://user-images.githubusercontent.com/48180714/232427119-2e930bea-40d8-4456-bb86-5871b2b5fc17.jpg" width="300">

**Comment by vvoovv** [comment]
> From the detection of the turn-lanes, the left and right widths of the ways at their start and their end have to be calculated so that the intersection or transition areas can be constructed, and their connectors set to the right place. The question now is how to pass this information to Blender. I suggest extending the attributes of the `TrimmedWaySection` class, but I don't know which method you prefer.

I think an offset from the centerline would be enough for rendering. So the name of the related attribute could be `surfaceCenterOffset`.

**Comment by vvoovv** [comment]
>  In reality, the case of a left turn in a bidirectional way usually just leads to a widening of the street, where both forward and backward lanes have to make slight curves (see an example in the right image). Therefore, my suggestion.

Can you sketch how it would look like?

**Comment by polarkernel** [comment]
> I think an offset from the centerline would be enough for rendering.

I don't think it's that simple. Let me explain the cases my framework is able to detect. All these cases are related to the direction of the centerline in the `TrimmedWaySection` class, which is the direction the centerline was drawn in OSM. Exactly where the center is, is a bit arbitrary and depends on the feeling the operator had when drawing the way. In the following sketches, the centerline always runs from left to right.

For `oneway` sections, the following cases are detected:

<img src="https://user-images.githubusercontent.com/48180714/232529972-3edcc410-d5cd-45ce-80bf-66c1adb5e1f1.png" width="500">

Note that the smaller part is on different ends, for _add_ and for _merge_. Currently, the case that two such events occurring in the same way-section, is not implemented, but could easily be added. For bidirectional ways, the situation is different:

<img src="https://user-images.githubusercontent.com/48180714/232532569-93ccbde5-7fa6-48f1-8c6a-64cbddebe85c.png" width="500">

So the offset from the centerline is not enough. You need to know the side (left or right) and the position of the smaller part (start or end of centerline). By the way: I don't understand, what you mean by "offset from the centerline", could you please sketch that?

> Can you sketch how it would look like?

In the current implementation, changes in section width due to changes in the number of lanes or changes in the way category were handled by an intersection area (internally in my code I called this a transition area). The result is shown in the upper image. Since it is difficult to handle a turn-left in a bidirectional way, I thought we could handle it the same way(lower image).

<img src="https://user-images.githubusercontent.com/48180714/232537598-92c83fcb-26c7-4cf8-be8c-b349b27b1386.png" width="500">

**Comment by vvoovv** [comment]
> Another ambiguous case. The turn lane detection can't know the side of the extra lanes from these tags:

Here is my suggestion. If it's a bidirectional way and there are doubts where to place extra lanes, place them to the right from the direction of movement. A bidirectional way does not suppose to have enough space for those extra lanes between the two directions of movements.

**Comment by vvoovv** [comment]
> Another ambiguous case. The turn lane detection can't know the side of the extra lanes from these tags:

I suggest adding all those test cases to our repo [osm_extracts](https://github.com/prochitecture/osm_extracts)/streets.

**Comment by vvoovv** [comment]
> <img alt="" width="100" src="https://user-images.githubusercontent.com/48180714/232537598-92c83fcb-26c7-4cf8-be8c-b349b27b1386.png">

I guess you didn't change the direction to the opposite one for the upper lane. Correct?

**Comment by polarkernel** [comment]
> I suggest adding all those test cases to our repo [osm_extracts](https://github.com/prochitecture/osm_extracts)/streets.

Currently, I only have the _json_ files I exported from overpass turbo. I developed the lane detection in a separate application for easier testing. As soon as I am done, I can search those lanes again and add the corresponding _osm_ files to this place.

**Comment by polarkernel** [comment]
> I guess you didn't change the direction to the opposite one for the upper lane. Correct?

Yes. Currently, for common ways, intersections of order two (two outgoing ways) are searched for and using the difference in width, a transition area is created (as shown in this sketch) and added to the `manager.intersectionAreas` list. For ways in clusters, a similar transition area is constructed, but this one additionally compensates for unequal widths at both ends of a cluster.

**Comment by vvoovv** [comment]
> So the offset from the centerline is not enough. You need to know the side (left or right) and the position of the smaller part (start or end of centerline). By the way: I don't understand, what you mean by "offset from the centerline", could you please sketch that?

Under the centerline I meant:
* for clusters: `WayCluster.centerline`
* for simple street sections: `TrimmedWaySection.centerline`

For one-way sections the attribute `TrimmedWaySection.offset` should be adjusted from its current value to take into account the transition lane(s).

To take into account the transition lanes, I suggest introducing the following attributes to the class `TrimmedWaySection`:
* transitionSL (a transition lane at the start of the section, on the left)
* transitionSR (a transition lane at the start of the section, on the right)
* transitionEL  (a transition lane at the end of the section, on the left)
* transitionEL  (a transition lane at the end of the section, on the left)

Those attributes will have the values False or True for now. Later the values can be expanded to a dedicated class.

**Comment by vvoovv** [comment]
> Do you agree, that for such bidirectional ways that add a turn way to the left, we just use a common transition and not a transition mesh?

Yes, I agree.

A separator line between the two directions of movements must be drawn. I hope it can accomplished by rotating the texture.

![image](https://user-images.githubusercontent.com/613295/232793483-79f766ff-a178-4404-874e-97ee753066e6.png)

**Comment by vvoovv** [comment]
In addition the following attributes are needed in the class `TrimmedWaySection`:
* trimLengthS (the trim length for the start of the centerline)
* trimLengthE (the trim length for the end of the centerline)

**Comment by polarkernel** [comment]
> Under the centerline I meant:
-for clusters: WayCluster.centerline
-for simple street sections: TrimmedWaySection.centerline
For one-way segments the attribute TrimmedWaySection.offset should be adjusted from its current value to take into account the transition lane(s).

I am still not sure, what you would like to have. Perhaps first some facts you may not be aware of. The centerline `TrimmedWaySection.centerline` follows the original way section (blue lines in the sketch below). In a cluster, the distance of a way from the cluster centerline (red line) may change between the start and the end of the cluster. The offset (red dotted lines), stored in `TrimmedWaySection.offset` is calculated as the average between these two distances. The cluster then has a constant width, based on these offsets and at the wider end of the cluster, a transition area corrects the missing part (not drawn). 

<img src="https://user-images.githubusercontent.com/48180714/232855432-ed221cb9-0cea-477d-a914-c62a02eaa9b4.png" width="300">

Now I am not sure how to adjust the `TrimmedWaySection.offset`. In the following picture I have drawn an example. The red line is the "centerline" given by the offsets as described above. Should it be on the centerline of the entire way, including the turn lane, as shown?

<img src="https://user-images.githubusercontent.com/48180714/232857078-51bcab0c-6e36-447e-ad0d-9f4760c324d1.png" width="500">

> To take into account the transition lanes, I suggest introducing the following attributes to the class TrimmedWaySection:
    -transitionSL (a transition lane at the start of the section, on the left)
    -transitionSR (a transition lane at the start of the section, on the right)
    -transitionEL (a transition lane at the end of the section, on the left)
    -transitionEL (a transition lane at the end of the section, on the left)
Those attributes will have the values False or True for now. Later the values can be expanded to a dedicated class.

OK.

**Comment by polarkernel** [comment]
> In addition the following attributes are needed in the class TrimmedWaySection:
    -trimLengthS (the trim length for the start of the centerline)
    -trimLengthE (the trim length for the end of the centerline)

The centerline of the `TrimmedWaySection` is already trimmed to fit between the intersection areas. Do you want to know how long these trimmed parts are? If so, be aware that these trimmed parts are not guaranteed to be straight.

**Comment by vvoovv** [comment]
I wanted to insert the following image to my [message](#issuecomment-1513147984) to illustrate what was meant under the _centerline_, but somehow forgot to do it.

![image](https://user-images.githubusercontent.com/613295/232861445-5fa9a96c-12f3-4927-9c5d-bfd98c70f9ce.png)

**Comment by vvoovv** [comment]
![image](https://user-images.githubusercontent.com/613295/232999512-00ee3522-db24-42b1-ab5d-d987133429f4.png)

The polylines I'd like to have are marked with green on the image above. They will be generated out of the centerline of the cluster using the offsets shown on the image (_offset 1_ and _offset 2_) and the trim lengths.

For the green polyline on the right the trim length for the end is equal to the length of the segment _2-3_ (the blue number at the bottom on the image).
For the  green polyline on the left the trim length for the start is equal to the length of the segment _1-2_.

**Comment by polarkernel** [comment]
Aha! Now I know where we are misunderstanding each other. In the general case, the transition (the red part in my drawing) may also be an intersection, with multiple outgoing ways or clusters. In my simple example, the two ways are two different instances of the `TrimmedWaySection` class. They might even be parts of different clusters. So you can't use the right segment for your construction.

The only data you get to construct a turn lane is the data of the left `TrimmedWaySection`. For the specific case in _berlin_karl_marx_allee.osm_, that you used to start this thread, their attributes could for instance look like this:
```
        self.centerline = [Vector((x,y)), ...]
        self.offset = 7.5
        self.width = 10.8
        self.nrOfLanes = 4
        self.transitionSL = True
        self.transitionSR = False
        self.transitionEL = False
        self.transitionER = False
```
`self.centerline` is the list of vertices of the trimmed original centerline from OSM. `self.offset` is the average offset, as depicted in the first image of my post [here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1513557026). If the way section is not part of a cluster, it is always zero. `self.width` is the width of the final four lanes (given by `self.nrOfLanes`).

Your new "offset", not to be confused with `self.offset`, would shift the "centerline" by half a lane width to the left.

There is also some confusion with the term _trim_. The _trim_ in `TrimmedWaySection` is used, because the length of `self.centerline` is a trimmed (shortened) version of the original centerline from OSM, so that it fits between the intersection areas.

**Comment by vvoovv** [comment]
> There is also some confusion with the term _trim_. The _trim_ in `TrimmedWaySection` is used, because the length of `self.centerline` is a trimmed (shortened) version of the original centerline from OSM, so that it fits between the intersection areas.

I've been using the term _trim_ due to the Blender's `Trim Curve` node.

![image](https://user-images.githubusercontent.com/613295/233175409-ac06a832-2f47-4346-ad6e-90b404a96d36.png)

![image](https://user-images.githubusercontent.com/613295/233178908-db7f40bf-72b4-4bcb-af7a-9e55d615e91e.png)

The red section represents the trim length at the start of the curve.
The blue section represents the trim length at the end of the curve.

**Comment by vvoovv** [comment]
> In the general case, the transition (the red part in my drawing) may also be an intersection, with multiple outgoing ways or clusters.

My original idea was to get rid of the transition polygons when the number of lanes is changed in a roadway. That would help to achieve a more realistic look of the roadway.

**Comment by polarkernel** [comment]
> I've been using the term trim due to the Blender's Trim Curve node.

OK, then I will change the term _trim_ on my side, maybe by _prune_ or _crop_.

**Comment by polarkernel** [comment]
> My original idea was to get rid of the transition polygons when the number of lanes is changed in a roadway. That would help to achieve a more realistic look of the roadway.

I see. The problem is, that both, `TrimmedWaySection` and `WayCluster`, map one single or multiple parallel ways, that do not change along the entire track, from end to end. I see two ways to overcome this problem, but both lead to a major redesign:

- We introduce a class similar to the existing `IntersectionArea` class, which connects two ways or two clusters without any space in between. This class, we could call it `GlueLine` or `AdhesiveLine`, would keep the two ends connected and could maybe hold some information, like your trim lengths. I call it Line, because all the connectors would be on one line.
- We introduce a new class, for example `WayTrack`, that connects two adjacent `TrimmedWaySections`, where one of them is the way with the turning lane. A cluster could then hold both, `TrimmedWaySections` and `WayTracks` at the same time.

**Comment by vvoovv** [comment]
I suggest combining both approaches.

Below are my thoughts.

Instances of the class `TrimmedWaySection` can be connected to instances of the classes:
* `IntersectionArea` (a real intersection with 3 or more street sections)
* `TransitionSymLane` (a symmetrical increase or decrease of the number of lanes)
* `TransitionSideLane` (an asymmetrical increase or decrease of the number of lanes when an additonal lane is attached to the left or/and to the right)

The class `TransitionSideLane` must be similar to the classes `GlueLine` or `AdhesiveLine` from you message. It contains information about transition lanes to the left and to the right.

**1. One-way street section (no cluster):**

`IntersectionArea` should be replaced for `TransitionSideLane` for the case below:
![image](https://user-images.githubusercontent.com/613295/233724889-1a5dea62-be6e-45ea-927c-846740c422a8.png)


The member `self.offset` of the class  `TrimmedWaySection` can be used to store the offset from the OSM-way for the way-section on the right.

`IntersectionArea` should be replaced for `TransitionSymLane` for the case below:
![image](https://user-images.githubusercontent.com/613295/233724975-f2a940f4-0ffc-4894-8bc7-ff4b9a8b27f0.png)


**2. Clustered street section**

A cluster consists of at least two roadways. Each roadway has one or more instances of the class  `TrimmedWaySection`. So the class `WayCluster` should have a list of roadways. Each roadway is a list of instances of the class `TrimmedWaySection` ordered in the direction of movement.
```python
self.roadways = [
    [TrimmedWaySection(..), TrimmedWaySection(..)],
    [TrimmedWaySection(..)],
    [TrimmedWaySection(..), TrimmedWaySection(..)]
]
```

Each instance of the class `TrimmedWaySection` should contain the trim lengths for the start and for the end of the instance. Those trim lengths are applied to the curve defined by `self.centerline`.

What if there is a partial intersection in the cluster? A roadway that is not a part of the intersection continues along its course. Instances of the class `TrimmedWaySection` will be connected to an `IntersectionArea` instead of `TransitionSideLane`.

**Comment by vvoovv** [comment]
**3. Bidirectional street section.**

Similar to the previous cases, two instances of the class `TrimmedWaySection` are attached to `TransitionSideLane` or `TransitionSymLane`.

However there should be no physical separator between the two directions of movements mapped by a single OSM-way. That means an additional lane (even if it's on the left) always leads to an increase of the width of the whole bidirectional roadway.

**Comment by vvoovv** [comment]
> That means an additional lane (even if it's on the left) always leads to an increase of the width of the whole bidirectional roadway.

There is an exception to this conclusion: an exchange of lanes between the two directions of movement.
![image](https://user-images.githubusercontent.com/613295/233846303-6ec09f15-2bf7-4f93-af14-712ae0597da7.png)

**Comment by polarkernel** [comment]
> I suggest combining both approaches.

First of all, I have a few questions to help me understand. In the following, I will replace the term `TrimmedWaySection` with `CroppedWaySection` to resolve the ambiguity with Blender's `Trim Curve` node.

> IntersectionArea should be replaced for TransitionSideLane for the case below:

Maybe I still do not understand your idea with the `TransitionSideLane`. Am I right in assuming that there will no more be any transition area between the `TransitionSideLane` and the preceding `CroppedWaySection`? The centerlines of the two`CroppedWaySections` are directly connected without any cropping to create one continuous curve. The offset attribute of the right `CroppedWaySection` stores the required offset from its centerline. For all common instances of `CroppedWaySection`, this value will be zero.

An instance of `TransitionSideLane` would need something like the following attributes:
```python
        self.connectedWayID = 123   # the ID of the smaller CroppedWaySection
        self.transitionWayID = 456  # the ID of the CroppedWaySection to be modified

        self.transitionSL = False   # define the type of the trsnsition
        self.transitionSR = True
        self.transitionEL = False
        self.transitionER = False

        self.trimStart = ???        # values for Blender's `Trim Curve` node,
        self.trimEnd = ???          # I still do not understand.
```
The direction of the centerlines of the two `CroppedWaySections` can be different, depending on the transition type (turn lane or merge lane).

> IntersectionArea should be replaced for TransitionSymLane for the case below:

But it will still be a polygon with the corresponding connectors, just like in the `IntersectionArea`? Or can it also be constructed in Blender?

> Clustered street section
Each roadway is a list of instances of the class TrimmedWaySection ordered in the direction of movement.

Let me illustrate my remarks using the (rotated) long cluster in _moscow_leningradski_prospekt_.osm:

<img src="https://user-images.githubusercontent.com/48180714/233946481-43eb7110-4f27-4052-8a43-a587de2a2ef1.jpg" width="700">

This will become one single cluster who's ways have the same offset from the cluster's centerline all along their entire length. At the intersection of one of its ends, a width transition is integrated to compensate for width differences at its ends.
To define these offsets, the suggested list of cluster ways
```python
self.roadways = [
    [CroppedWaySection(..), CroppedWaySection(..)],
    [CroppedWaySection(..)],
    [CroppedWaySection(..), CroppedWaySection(..)]
]
```
could be accompanied by a list of offsets to the cluster's centerline in the same order:
```python
self.offsets= [ -10.5, -1.2, 9.8 ]
```
Note that the centerlines of the `CroppedWaySections` may be at different positions than these offsets. This is also visible in the image above. Also the direction of the centerline in `CroppedWaySection` will be the original direction, as it has been drawn, and not the direction of the cluster's centerline.

> What if there is a partial intersection in the cluster? A roadway that is not a part of the intersection continues along its course. Instances of the class TrimmedWaySection will be connected to an IntersectionArea instead of TransitionSideLane.

I agree. The `IntersectionArea` and its connectors for such a case will be constructed using the offsets above and not using the centerlines of the `CroppedWaySections` in the list above.

> However there should be no physical separator between the two directions of movements mapped by a single OSM-way. That means an additional lane (even if it's on the left) always leads to an increase of the width of the whole bidirectional roadway.

Could you please sketch this idea, I am not sure if I understand it?

> There is an exception to this conclusion: an exchange of lanes between the two directions of movement.

You are right. I'm afraid this case is not even detected in the current version.

**Comment by vvoovv** [comment]
> I will replace the term `TrimmedWaySection` with `CroppedWaySection` to resolve the ambiguity with Blender's `Trim Curve` node.

Can we get rid of the prefix _Trimmed_ or _Cropped_ at all. For example, it could be renamed to `RoadwaySection` or `StreetSection`.

> Maybe I still do not understand your idea with the `TransitionSideLane`. Am I right in assuming that there will no more be any transition area between the `TransitionSideLane` and the preceding `CroppedWaySection`? The centerlines of the two`CroppedWaySections` are directly connected without any cropping to create one continuous curve. The offset attribute of the right `CroppedWaySection` stores the required offset from its centerline. For all common instances of `CroppedWaySection`, this value will be zero.

Yes, everything is correct.

> ```python
>         self.connectedWayID = 123   # the ID of the smaller CroppedWaySection
>         self.transitionWayID = 456  # the ID of the CroppedWaySection to be modified
> ```

I posted an example of the exchange of lanes in the previous [message](#issuecomment-1519083445). In that case it isn't easy to assign `self.connectedWayID` and `self.transitionWayID`. That's why I suggest using

```python
self.ways = (123, 456)
```

where `self.ways[0]` is an id of the first way along the original OSM direction and `self.ways[1]` is an id of the second one.
> ```python
>         self.transitionSL = False   # define the type of the transition
>         self.transitionSR = True
>         self.transitionEL = False
>         self.transitionER = False
> ```

Those attributes are not relevant anymore. For now I suggest to introduce at least two attributes that indicate if there is a transition lane to the left and to the the right from the narrower way-section:
```python
self.laneR = True
self.laneL = False
```

In the case of a bidirectional way-section, the values of those attributes can be tuples with two elements.

> ```python
>         self.trimStart = ???        # values for Blender's `Trim Curve` node,
>         self.trimEnd = ???          # I still do not understand.
> ```

The attributes above are not needed for `TransitionSideLane`. They are needed for `CroppedWaySection` and only if it is part of a cluster.

> But it will still be a polygon with the corresponding connectors, just like in the `IntersectionArea`? Or can it also be constructed in Blender?

A polygon will be needed. It can't be constructed in Blender.

> Also the direction of the centerline in `CroppedWaySection` will be the original direction, as it has been drawn, and not the direction of the cluster's centerline.

Currently an attribute describing the direction of a cluster part is missing. It can be called `self.alongCenterline`.

> Could you please sketch this idea, I am not sure if I understand it?

I'll sketch it in the next message.

**Comment by vvoovv** [comment]
> However there should be no physical separator between the two directions of movements mapped by a single OSM-way. That means an additional lane (even if it's on the left) always leads to an increase of the width of the whole bidirectional roadway.

There is an agreement in OpenStreetMap, that if there is a physical separator (grass, hatched area, all kinds of traffic separator) between the roadways in the opposite direction, then the roadways should be mapped by separate OSM ways. Otherwise a single OSM way can be used.
![image](https://user-images.githubusercontent.com/613295/234275637-e94ffaaf-3e00-4693-93bb-0166e73e1d1d.png)
<img src="https://user-images.githubusercontent.com/48180714/232226999-222431fd-0525-426e-897f-65088439d6a6.jpg" width="300">

If there is a turning lane to the left, it can be placed on the separator area without increasing the overall width of the cluster:

![image](https://user-images.githubusercontent.com/613295/234280669-42b52cea-b33c-434b-8dcd-fe6820be6d18.png)

In the case of a bidirectional OSM way, there is no space to place a turning lane to the left. That leads to a configuration likes the one below:
![image](https://user-images.githubusercontent.com/613295/232793483-79f766ff-a178-4404-874e-97ee753066e6.png)

**Comment by polarkernel** [comment]
> Can we get rid of the prefix _Trimmed_ or _Cropped_ at all. For example, it could be renamed to `RoadwaySection` or `StreetSection`.

We can do that, I would prefer `StreetSection`. In `StreetGenerator` I use a `WaySection` class that holds the original OSM way segments and, beside many other attributes,  a reference to the section network graph. The centerline in `StreetSection` is truncated to fit between the intersection areas, therefore I used the term _trimmed_.

> I posted an example of the exchange of lanes in the previous [message](#issuecomment-1519083445). In that case it isn't easy to assign `self.connectedWayID` and `self.transitionWayID`. That's why I suggest using `self.ways = (123, 456)`
where `self.ways[0]` is an id of the first way along the original OSM direction and `self.ways[1]` is an id of the second one.

OK. An alternative would be to use the same rules as for the connectors in `IntersectionArea`, positive, when the start of the way connects to `TransitionSideLane`, and negative, if it ends there. Then the order in the tuple would be arbitrary.
```python
self.ways = (-123, 456)
```
> Those attributes are not relevant anymore. For now I suggest to introduce at least two attributes that indicate if there is a transition lane to the left and to the the right from the narrower way-section:
self.laneR = True
self.laneL = False

OK. So when I summarize the above in an example, assuming that it was drawn from left to right, we would get these attributes:

<img src="https://user-images.githubusercontent.com/48180714/234527364-1250862c-ca60-42d1-84b5-7bce719e6021.png" width="750">

> self.trimStart = ???        # values for Blender's `Trim Curve` node,
self.trimEnd = ???          # I still do not understand.
The attributes above are not needed for TransitionSideLane. They are needed for CroppedWaySection and only if it is part of a cluster.

I still do not understand these attributes. Why only in a cluster?

> Currently an attribute describing the direction of a cluster part is missing. It can be called `self.alongCenterline`.

OK. Should we possibly consider some internationalization, to handle the left-hand traffic case?

> There is an agreement in OpenStreetMap, that if there is a physical separator (grass, hatched area, all kinds of traffic separator) between the roadways in the opposite direction, then the roadways should be mapped by separate OSM ways. Otherwise a single OSM way can be used.

In the lane detector, I used the `oneway` tag to distinguish this case. If mapped as separate ways, the ways have to be declared as one-way, don't they?

**Comment by polarkernel** [comment]
>vvoovv: What if there is a partial intersection in the cluster? A roadway that is not a part of the intersection continues along its course. Instances of the class `TrimmedWaySection` will be connected to an `IntersectionArea` instead of `TransitionSideLane`.

>polarkernel: I agree. The `IntersectionArea` and its connectors for such a case will be constructed using the offsets above and not using the centerlines of the `CroppedWaySections` in the list above.

There is an issue I like to discuss:

<img src="https://user-images.githubusercontent.com/48180714/234584115-be24562f-d6d6-4419-b544-3065a30c6143.png" width="650">

Suppose there are two outermost consecutive ways in the list `self.roadways` of a cluster, as shown in 1). Their offset to the centerline (red) of the cluster is given in the list `self.offsets`, and their width in their instances of `StreetSection`. Since this offset is constant along the entire length of the cluster, the partial intersection with an outgoing way can lie anywhere outside this offset, as shown in 2). However, using the centerline of the cluster, I can construct the instance of `IntersectionArea`, shown as the red area in 3), with three connectors in its dictionary `self.connectors`.

However, the two GN ways in the cluster now need to be shortened accordingly. Shortening the original centerlines for the `StreetSection` instances, as I do for the out-of-cluster ways, does not help, because they are different. I don't know how you construct GN ways, can you possibly solve this problem using the connectors?

**Comment by vvoovv** [comment]
> I would prefer `StreetSection`.

Ok. Let's rename it to `StreetSection`.

> OK. An alternative would be to use the same rules as for the connectors in `IntersectionArea`, positive, when the start of the way connects to `TransitionSideLane`, and negative, if it ends there. Then the order in the tuple would be arbitrary.

Let's do it the way you proposed for consistency.

> OK. So when I summarize the above in an example, assuming that it was drawn from left to right, we would get these attributes:

Yes. Everything seems to be correct on the drawing.

> I still do not understand these attributes. Why only in a cluster?

To render a cluster the following actions are made for each part of the cluster:
* the centerline of the cluster is trimmed from the start and from the end, if `StreetSection` does not span the whole course of the centerline
* an offset is applied to the resulting polyline

I assumed that in the case of a simple non-clustered `StreetSection` there is no need to trim `StreetSection.centerline` since it spans the whole course of `StreetSection.centerline`.


> OK. Should we possibly consider some internationalization, to handle the left-hand traffic case?

Yes. The addon is used in UK quite actively.

> In the lane detector, I used the `oneway` tag to distinguish this case. If mapped as separate ways, the ways have to be declared as one-way, don't they?

Yes.

**Comment by vvoovv** [comment]
> I don't know how you construct GN ways, can you possibly solve this problem using the connectors?

I need `self.trimStart` and `self.trimEnd` to get the resulting street sections out the centerline of the cluster.

**Comment by polarkernel** [comment]
> I need self.trimStart and self.trimEnd to get the resulting street sections out the centerline of the cluster.

OK. But then I assume that the GN ways are exactly parallel to the cluster centerline. This is not always the case for an extended curved polyline. I presume the dimensions of `self.trimStart` and `self.trimEnd` are in meters.

In most cases, the ways at the ends of a cluster are truncated by the intersection area. I will use these values there as well.

**Comment by vvoovv** [comment]
> But then I assume that the GN ways are exactly parallel to the cluster centerline.

Yes, that is assumed.

> I presume the dimensions of `self.trimStart` and `self.trimEnd` are in meters.

Yes, they are in meters.

**Comment by vvoovv** [comment]
Another option that I can explore for clusters.

Instead of generating the centerline of each `StreetSection` from `WayCluster.centerline`, the original OSM way stored at `StreetSection.centerline` can be used for that.

I can explore if it's possible to generate the separators between roadways.

**Comment by polarkernel** [comment]
> I can explore if it's possible to generate the separators between roadways.

Currently, this doesn't work. The `createWaySection` function doesn't even fill in the `StreetSection.centerline` attribute, so I didn't truncate it either.

In order to implement all the changes resulting from this thread, all the following methods will need to be completely reworked to some degree:
```
        self.detectWayClusters()
        self.createLongClusterWays()
        self.createClusterTransitionAreas()
        self.createClusterIntersections()
        self.createClippedClusterEnds()
        self.createWayClusters()
        self.cleanCoveredWays()
        self.createOutput()
```
We have a major refactoring ahead of us. As you can see, the construction of the clusters has to be done first. So I suggest that we try to find a solution to this problem as soon as the first clusters are available.

**Comment by vvoovv** [comment]
> I can explore if it's possible to generate the separators between roadways.

It was a proposal for an alternative approach how to render clusters. 

> We have a major refactoring ahead of us. 

Then the alternative approach has to wait till the the refactoring of the existing approach is finished.

**Comment by polarkernel** [comment]
How should we handle the lane in the middle of way [473625784](https://www.openstreetmap.org/way/473625784)?

<img src="https://user-images.githubusercontent.com/48180714/236210079-4d050d80-e60c-4ed8-ab5d-9c739fe41c96.jpg" width="500">

The tags that are specific to this lane are:
```
lanes:both_ways=1
turn:lanes:both_ways=left
```
Just add as a forward lane?

**Comment by vvoovv** [comment]
> How should we handle the lane in the middle of way [473625784](https://www.openstreetmap.org/way/473625784)?

Will it disappear as a part of the intersection polygon?

**Comment by polarkernel** [comment]
> Will it disappear as a part of the intersection polygon?

No, in this case not:

<img src="https://user-images.githubusercontent.com/48180714/236398159-e0969464-1dce-4ec6-ad27-efc5f147972b.jpg" width="250">

These are usually short ways, but in many cases, they will appear in the results. A few more examples: [903857688](https://www.openstreetmap.org/way/903857688), [686964216](https://www.openstreetmap.org/way/686964216), [485351420](https://www.openstreetmap.org/way/485351420), [456793245](https://www.openstreetmap.org/way/456793245), [733883679](https://www.openstreetmap.org/way/733883679), [940905385](https://www.openstreetmap.org/way/940905385), 

This tag seems to be distributed rather unevenly. For example, while I found only two of them in Berlin, I found 34 in Hamburg. Anyway, I would like to have a general solution for this.

**Comment by vvoovv** [comment]
> ```
> lanes:both_ways=1
> ```

This looks like a special case. There are 62546 of them in the OSM database. So I suggest to treat like a special case.

Each street section would contain then:
* forward lanes
* backward lanes
* lanes with both directions

**Comment by polarkernel** [comment]
> This looks like a special case. There are 62546 of them in the OSM database. So I suggest to treat like a special case.
Each street section would contain then:

OK. The relevant attributes of my example way [473625784](https://www.openstreetmap.org/way/473625784), tagged as:
```
    lanes = 3
    lanes:forward = 1
    lanes:backward = 1
    lanes:both_ways = 1
    turn:lanes:forward = through
    turn:lanes:backward = through
    turn:lanes:both_ways = left
```
would then become
```python
    self.nrOfLanes = (1,1,1)
    self.laneR = (False,False)
    self.laneL = (False,False)
```
where the last number in the tuple `nrOfLanes` is for both directions. For `laneR` and `laneL` it makes no sense to append a value for the `both_ways` lane. Its value will always be `left` for left-hand traffic and `right` else.

**Comment by polarkernel** [comment]
I am currently considering how to represent left and right traffic in StreetSection. Let me discuss the problems using a simple example depicted below. I assume the red arrows as forward (drawn) direction of the way-section in OSM.

<img src="https://user-images.githubusercontent.com/48180714/236473244-f1cce06b-60c7-4b16-8e95-b99ccb69ba32.png" width="600">

How can we represent left- and right-hand traffic? Currently, I use an argument `leftHandTraffic` in the function `__init__()` of the `StreetGenerator` class, so that the addon can tell to my code, which version to use. The attributes of these two way-sections in the `StreetSection` class are shown below. But they do not distinguish between left- and right-hand traffic.

```python
    # left-handed                                   # right-handed
    self.nrOfLanes = (2,1,0)                        self.nrOfLanes = (2,1,0)
    self.laneR = (True,False)                       self.laneR = (False,False)
    self.laneL = (False,False)                      self.laneL = (True,False)
```
Will the addon take care of it?

A second problem are the turn directions of the backward lanes. Suppose I have a way with the following tags in a scene that uses right-hand traffic:
```
    lanes = 3
    lanes:forward = 1
    lanes:backward = 2
    turn:lanes:forward = through
    turn:lanes:backward = through|right
```
Then the turn lane goes to the right, seen in backward direction. Am I right for this case, when I set `self.laneR = (False,True)`?

**Comment by vvoovv** [comment]
> ```python
>     self.nrOfLanes = (1,1,1)
> ```

I suggest to represent that similar to OSM:
> ```python
>     self.lanesForward = 1
>     self.lanesBackward = 1
>     self.lanesBoth = 1
> ```

The default value for `self.lanesBoth` set in the constructor is zero. Then there is no need to change it later in the code if there is no lane with both directions.

**Comment by polarkernel** [comment]
> I suggest to represent that similar to OSM:

OK, I like this more than the version we had for TrimmedWaySection, which I have extended my suggestion above. For `oneway` ways, we can just set `self.lanesBackward = 0`.

**Comment by vvoovv** [comment]
> Will the addon take care of it?

Yes, there will a global variable `app.rht` (_True_ for the right-hand traffic and _False_ for the left-hand traffic). Not sure though if it's needed for the class `StreetSection`. For now `app.rht` will be set in the GUI panel.

> Am I right for this case, when I set `self.laneR = (False,True)`?

Yes.

**Comment by vvoovv** [comment]
At the moment I'd like to implement the simplest solution. For the seamless transitions between the lanes, the ratio of the widths of the roadways with 3 and 2 lanes must be equal to 676/448=1.5089. This ratio is defined by the width of the related textures.

![image](https://github.com/prochitecture/blosm/assets/613295/77d7d03b-643a-47ac-9eca-27f7278b5ebd)


If it is ok, I can provide the ratios for the other pairs of number of lanes.

**Comment by polarkernel** [comment]
> If it is ok, I can provide the ratios for the other pairs of number of lanes.

If I understand you correctly, this means that the width of the transition section (3 lanes) is 1.5089 times the width of the two lanes of the preceding section, and the `offset` given in the wider section is adjusted accordingly. This should be feasible, since the transition section usually ends in an intersection (where its connector can be adjusted) and not in a continuing way.

Do these widths include the value `'doubleRoadsideWidth'` given in `wayCategoryProps` ([line 7](https://github.com/prochitecture/blosm/blob/b2199eea30b9a8a8f5e63a7a985ca618ecba6151/way/way_properties.py#LL7C1-L7C17) in _way/way_properties.py_)?

If I am right, then please provide the other ratios.

**Comment by polarkernel** [comment]
In [this comment](https://github.com/prochitecture/blosm/issues/57#issuecomment-1519083445), you mentioned the exchange of lanes between the two directions of movement:

![image](https://user-images.githubusercontent.com/613295/233846303-6ec09f15-2bf7-4f93-af14-712ae0597da7.png)

Is this a real case and, if yes, how is it tagged?

**Comment by vvoovv** [comment]
> If I understand you correctly, this means that the width of the transition section (3 lanes) is 1.5089 times the width of the two lanes of the preceding section, and the `offset` given in the wider section is adjusted accordingly. 

Yes, that's correct.

> Do these widths include the value `'doubleRoadsideWidth'` given in `wayCategoryProps` ([line 7](https://github.com/prochitecture/blosm/blob/b2199eea30b9a8a8f5e63a7a985ca618ecba6151/way/way_properties.py#LL7C1-L7C17) in _way/way_properties.py_)?

Roadways in cities and towns rarely have a roadside. That's why I suggest omitting it.

> If I am right, then please provide the other ratios.

2_lanes/1_lanes = 448/220 = 2.03636
3_lanes/2_lanes = 676/448
4_lanes/3_lanes = 904/676

A roadway width in pixels can be calculated as 220&times;Lanes+8&times;Lines, where _Lanes_ is the total number of lanes and _Lines_ is the total number of marking lines. If the two directions are separated by a double solid line, then 16 pixels are added  to the total width of the roadway.

**Comment by polarkernel** [comment]
> Roadways in cities and towns rarely have a roadside. That's why I suggest omitting it.

OK, I will omit it. This will solve several way-overlap issues. I will leave the values in the table, in case we need them later.

> A roadway width in pixels can be calculated as 220×Lanes+8×Lines, where Lanes is the total number of lanes and Lines is the total number of marking lines. If the two directions are separated by a double solid line, then 16 pixels are added to the total width of the roadway.

I will integrate this calculation into the code. Do we already have cases with a double solid line, and if so, how is this detected?

**Comment by vvoovv** [comment]
> Do we already have cases with a double solid line, and if so, how is this detected?

I can only propose some heuristics. Use a double solid line between the two directions of traffic if
* OSM-way does not have the tag `oneway=yes` (i.e. there is a traffic in both directions)
* The total number of lanes is equal or greater than 4.

I think the class `StreetSection` needs an attribute with the proposed name `separatorType` and possible values:

```python
class SeparatorType:
    unknown = 0
    dashedLine = 1
    solidLine = 2
    doubleSolidLine = 3
```

In the future a more advance renderer can be implemented which will be able to process totally arbitrary lane widths.

**Comment by polarkernel** [comment]
> I can only propose some heuristics.

I suggest that we hold off on that for now. I don't want to overload the individual development steps.

**Comment by vvoovv** [comment]
> I suggest that we hold off on that for now. I don't want to overload the individual development steps.

Ok

**Comment by vvoovv** [comment]
> In [this comment](https://github.com/prochitecture/blosm/issues/57#issuecomment-1519083445), you mentioned the exchange of lanes between the two directions of movement:
> Is this a real case and, if yes, how is it tagged?

The real example of a lane exchange is [attached](https://github.com/prochitecture/blosm/files/11455688/piestany.zip).

![image](https://github.com/prochitecture/blosm/assets/613295/3b897955-dd70-4bc1-9b8c-4fbf00fa25e0)
I rotated the above image 90 degrees to save the space.

It's mapped as it's supposed to be mapped:
```
highway = primary
lanes = 3
lanes:forward = 1
lanes:backward = 2
```
and
```
highway = primary
lanes = 3
lanes:forward = 2
lanes:backward = 1
```

**Comment by vvoovv** [comment]
How should the transition lane be positioned relative to the OSM-node where the number of lanes is changed?

(A)
![image](https://github.com/prochitecture/blosm/assets/613295/1dd69d41-7b68-4ec5-a957-2d349856f141)

(B)
![image](https://github.com/prochitecture/blosm/assets/613295/69900b76-f103-4541-ad6e-fb870bb8f2bf)

(C)
![image](https://github.com/prochitecture/blosm/assets/613295/208e1f56-7fc9-42f1-a360-7d8c0174b434)

**Comment by polarkernel** [comment]
> How should the transition lane be positioned relative to the OSM-node where the number of lanes is changed?

I would suggest using (A). In most cases, it will best match the sections drawn by the OSM volunteers. They are usually set quite realistic (the red dots are the OSM nodes):

<img src="https://github.com/prochitecture/blosm/assets/48180714/499975e5-a1fc-44e3-8dec-f5365b5ad7f9" width="450">

**Comment by vvoovv** [comment]
But the transition node on the left is located at the middle of the transition lane. I am referring to the example in your previous [message](#issuecomment-1546633603).

**Comment by polarkernel** [comment]
> But the transition node on the left is located at the middle of the transition lane. I am referring to the example in your previous [message](https://github.com/prochitecture/blosm/issues/57#issuecomment-1546633603).

I don't think so. In reality, the transition itself is just steeper than on the right side, but the node is on the correct place (more or less):

<img src="https://github.com/prochitecture/blosm/assets/48180714/783f5884-5f30-4aa4-bdb1-a397ad529578" width="450">

**Comment by vvoovv** [comment]
At least (A) and (B) can be generated similarly. (C) is different. It would require to generate a triangle area.

**Comment by polarkernel** [comment]
> At least (A) and (B) can be generated similarly.

In this case, it is maybe best to search for the best solution experimentally, once I am able to deliver the first part of my development.

**Comment by polarkernel** [comment]
Finally, made it! A first version with `transitionSideLanes` and `transitionSymLanes` is committed.

There are two new classes, `TransitionSideLane` and `TransitionSymLane`, according to the previous discussion. Their attributes are described in the comments. Their instances are accessible by two new Python lists in manager:
```Python
        manager.transitionSideLanes
        manager.transitionSymLanes
```
The `StreetSection` class has been extended by the attributes `forwardLanes`, `backwardLanes`, `bothLanes`, `laneR` and `laneR`. Its new `trimStart` and `trimEnd` attributes are not yet used.

The code does not yet provide clusters, and it has not yet been tested for right-handed traffic. The roadsides of the ways are now omitted. The [pixel ratio corrections](https://github.com/prochitecture/blosm/issues/57#issuecomment-1544025403) for additional lanes are implemented (and will hopefully work). The code for the intersection areas is still improvised from the old code. It will be replaced later, maybe a solution for overlapping ways can once be found.

I hope this version will help you do your first tests on transition lanes.

**Comment by vvoovv** [comment]
It would be good to access the related instances of `TransitionSideLane`, `TransitionSymLane`, `IntersectionArea` within an instance of the class `StreetSection`.

Currently the class `StreetSection` has the attributes `startConnected` and `endConnected`. I suggest to replace them for the attributes `start` and `end` respectively. They would be set to the related instanced of `TransitionSideLane`, `TransitionSymLane`, `IntersectionArea` or none if the start or the end of the street section is a dead-end. Later a dedicated class for a dead-end can be introduced.

I hope that won't require a big refactoring.

**Comment by vvoovv** [comment]
> It would be good to access the related instances of `TransitionSideLane`, `TransitionSymLane`, `IntersectionArea` within an instance of the class `StreetSection`.

I temporarily implemented that in the renderer by iterating through `manager.transitionSideLanes`. But I think it would be more natural to have the code in the generator since an instance of the class `StreetSection` should have access their to the related instances of `TransitionSideLane`, `TransitionSymLane`, `IntersectionArea`.

**Comment by vvoovv** [comment]
> ```python
> self.roadways = [
>     [TrimmedWaySection(..), TrimmedWaySection(..)],
>     [TrimmedWaySection(..)],
>     [TrimmedWaySection(..), TrimmedWaySection(..)]
> ]
> ```

I'd like to note that all instances of `TrimmedWaySection` that constitute a roadway, may be recovered from a single instance if an instance of `TrimmedWaySection` contains the attributes `start` and `end` as described in the message [above](#issuecomment-1550232423).

Later after some test I'll post a message about the structure of `self.roadways`.

**Comment by vvoovv** [comment]
> Later after some test I'll post a message about the structure of `self.roadways`.

The original proposed way of the structure of `self.roadways` is preferred:

```python
self.roadways = [
    [TrimmedWaySection(..), TrimmedWaySection(..)],
    [TrimmedWaySection(..)],
    [TrimmedWaySection(..), TrimmedWaySection(..)]
]
```

I am generating a list of instances of `TrimmedWaySection` for a non-clustered roadway in the renderer.

**Comment by vvoovv** [comment]
> I temporarily implemented that in the renderer by iterating through `manager.transitionSideLanes`. But I think it would be more natural to have the code in the generator since an instance of the class `StreetSection` should have access their to the related instances of `TransitionSideLane`, `TransitionSymLane`, `IntersectionArea`.

It's currently done in the temporary method [tmpPrepare](https://github.com/prochitecture/blosm/blob/dev/way/renderer_streets.py#L799) of the class `StreetRenderer`.

**Comment by polarkernel** [comment]
> It's currently done in the temporary method [tmpPrepare](https://github.com/prochitecture/blosm/blob/dev/way/renderer_streets.py#L799) of the class StreetRenderer.

I will implement the following changes for the branch 'streets_for_gn' (`tmpPrepare()` does not exist in this branch):
- `startConnected `and `endConnected` of the class `StreetSection` will be replaced by `start` and `end` and will hold an instance of either `TransitionSideLane`, `TransitionSymLane` or `IntersectionArea`, if connected to one of those, or is `None` else.
- The classes `TransitionSideLane` and `TransitionSymLane` get the new attributes `incoming` and `outgoing`, which hold an instance of the connected `StreetSection`.

If one of the following attributes, currently set in your temporary code `tmpPrepare()` should be set by `generate_streets`, please tell me: In `StreetSection`: `streetSectionIndex`, `numPoints`, `totalLanes`, `rendered`, `chunkedRoadway`, and in `TransitionSideLane`: `length`, `lengthOffset`, `totalLanesIncreased`.

**Comment by vvoovv** [comment]
> I will implement the following changes for the branch 'streets_for_gn' (`tmpPrepare()` does not exist in this branch):

Ok.

I've just merged the branch `dev` into `streets_for_gn`.

**Comment by vvoovv** [comment]
**StreetSection**
* `streetSectionIndex` will be set in the renderer
* `numPoints` should be set in `generate_streets` to `len(streetSection.centerline)`
* `totalLanes` should be set in `generate_streets` to `streetSection.forwardLanes + streetSection.backwardLanes + streetSection.bothLanes`
* `rendered` should be set in `generate_streets` to `False`
* `chunkedRoadway` will be described in the next message

**TransitionSideLane**
* `length` will be set in the renderer
* `lengthOffset` will be set in the renderer
* `totalLanesIncreased` should be set in `generate_streets` to `transition.outgoing.totalLanes > transition.incoming.totalLanes`

**Comment by vvoovv** [comment]
> * `chunkedRoadway` will be described in the next message

For now it will be set in the renderer.

**Comment by polarkernel** [comment]
I committed an intermediate development step to the branch 'streets_for_gn':

<img src="https://github.com/prochitecture/blosm/assets/48180714/5ea4bac0-a61a-4509-ab75-93383f3b3acb" width="500">

There are still issues due to conflicts of intersections for some datasets, but the data structure is already as desired. But I didn't yet touch _tmpPrepare()_.

**Comment by vvoovv** [comment]
I cleaned up `tmpPrepare()` in the class `StreetRenderer` from the code that is done now in `generate_streets`.

**Comment by polarkernel** [comment]
> I cleaned up tmpPrepare() in the class StreetRenderer from the code that is done now in generate_streets.

Thanks. I will try now to solve the issues by intersection conflicts. Maybe I will have to rewrite this code, because I do no more understand the code relicts I left in May. However, many scenes work without crashes, so that you may check whether the results are already useful. Later, I will continue by the implementation of the cluster ways.

**Comment by vvoovv** [comment]
A good plan!

There is still a problem #64 

What scenes lead to a crash?

I am trying to figure out how to describe the style of streets with the PML syntax.

**Comment by polarkernel** [comment]
> What scenes lead to a crash?

Currently, it is only in _osm_extracts/streets/milano_01.osm_. But there may be bugs in intersection areas. There is still the well-known old problem of "short ways".

**Comment by vvoovv** [comment]
I am getting the following error with _osm_extracts/streets/milano_01.osm_:
```python
  File "blosm\action\generate_streets.py", line 373, in do
    self.createOutput()
  File "blosm\action\generate_streets.py", line 1693, in createOutput
    streetSection = self.waySectionLines[abs(streetSectionIdx)]
KeyError: 192
```

**Comment by polarkernel** [comment]
>  am getting the following error with osm_extracts/streets/milano_01.osm:

Yes, that's what I meant by crash. The other scenes run to the end without error message.

**Comment by polarkernel** [comment]
I committed a version, that merges overlapping intersection areas and removes unresolvable overlapping ways (see issue #65) from this computation. Now, all scenes should run without crashing, but maybe still with some minor issues.

**Comment by vvoovv** [comment]
Thanks. Is it possible also to fix #64?

**Comment by polarkernel** [comment]
I am currently constructing the OSM files of all the special cases we found in this thread. I will add all those test cases to our repo [osm_extracts](https://github.com/prochitecture/osm_extracts)/streets. In the process, I've come across further bugs in the construction of turn lanes, that I like to fix before I continue with the clusters.

**Comment by polarkernel** [comment]
There is a problem with using the `offset` attribute in the `StreetSection` class when we have turn lanes. Until now, the direction of the centerlines of the ways has been the same as drawn by the OSM operator. But among the test cases mentioned in the last post, there was an example where both centerlines ran towards the transition node. 

The question now is how to define `offset`. One possibility is to use the direction from the transition node to the end of the wider path and define left as negative and right as positive offset, for instance. The other possibility is to use the direction of the centerline as reference. How does the add-on define this attribute?

**Comment by vvoovv** [comment]
> But among the test cases mentioned in the last post, there was an example where both centerlines ran towards the transition node.

Could you please point to this node?

**Comment by polarkernel** [comment]
> Could you please point to this node?

I was wrong, in the original OSM data my assumption is not true. The directions seem to be reversed somewhere in the process of `StreetGenerator`.  But the question remains in my opinion. The OSM operator could have drawn the directions from the smaller section to the section with the turn lane, or vice versa. What should be the reference for `offset`? What should it be for two-way streets?

**Comment by vvoovv** [comment]
> The question now is how to define `offset`. One possibility is to use the direction from the transition node to the end of the wider path and define left as negative and right as positive offset, for instance. The other possibility is to use the direction of the centerline as reference. How does the add-on define this attribute?

The renderer applies offset to a Blender curve relative to the supplied centerline. If the offset is negative, the curve is offseted to the left relative to the supplied centerline. Otherwise the curve is offseted to the right.

**Comment by polarkernel** [comment]
> The renderer applies offset to a Blender curve relative to the supplied centerline. If the offset is negative, the curve is offseted to the left relative to the supplied centerline. Otherwise the curve is offseted to the right.

OK, thanks. Let's lock this down so that the renderer is served correctly.  The reference is the centerline of the section with the turn lane (which is always the `outgoing` section of the `TransitionSideLane`, even if the centerline points inwards). If the offset is negative, the curve is offset to the left relative to the supplied centerline. Otherwise, the curve is offset to the right.

**Comment by vvoovv** [comment]
I proposed in #66 how to define the direction of a cluster or a two-way roadway.

**Comment by polarkernel** [comment]
I have created OSM files for all the special cases discussed in this thread and added these test cases to our [osm_extracts](https://github.com/prochitecture/osm_extracts)//streets repo. Many bugs have been found using them and their fixes have been committed. The files in detail are

_alsenstrasse.osm_, [discussed here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1494306695)
As expected, a `transitionSymLane` is constructed.

_nabrezie.osm_, [discussed here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1494383245)
Although there are two tags with turns to the left, the way remains a standard way, which is correct.

_canal_street.osm_, [discussed here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1509845657) (first case)
The turn lane is correctly constructed to the right.

_versmannstrasse.osm_, [discussed here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1509845657) (second case)
The turn lane is correctly constructed to the left.

_container_terminal_hamburg.osm_, [discussed here](https://github.com/prochitecture/blosm/issues/57#issuecomment-1510384370)
Additional turn lane correctly to the right, but only one lane instead of two.

**Comment by polarkernel** [comment]
>  Let's lock this down so that the renderer is served correctly.

Fixed and committed.
